R/Step2_QUAIL_vQTL.R

```diff
@@ -0,0 +1,199 @@
+                        g = .lm.fit(cbind(1, X[,i]), y)
+                        return(c(coef(g)[2], calc_stderr(cbind(1, X[,i]), g$residuals)[2]))
+                    })), silent = TRUE)
+
+# Exception handling
+if (inherits(output, "try-error")) {
+    output = matrix(0, ncol(X), 2)
+    for (i in 1:ncol(X)) {
+    fit = summary(lm(y ~ X[,i]))$coef
+    if (nrow(fit) == 2)
+        output[i,] = as.vector(summary(lm(y ~ X[,i]))$coef[2,1:2])
+    else
+        output[i,] = c(0,0)
+    }
+}
+
+# Calculate z-scores (t-statistics)
+z_scores = output[,1] / output[,2]
+
+# Calculate p-values from z-scores
+p_values = 2 * pnorm(abs(z_scores), lower.tail = FALSE)
+
+# Calculate q-values using the provided function
+q_values = compute_qvalues(p_values)
+
+# Use residual_x's column names as the column names for the results
+rownames(output) <- colnames(residual_x)
+
+result_list = list(
+    betahat = setNames(output[,1], rownames(output)),
+    sebetahat = setNames(output[,2], rownames(output)),
+    z_scores = setNames(z_scores, rownames(output)),
+    p_values = setNames(p_values, rownames(output)),
+    q_values = setNames(q_values, rownames(output))
+)
+
+if (return_residuals && !is.null(Z)) {
+    result_list$residuals = y
+}
+
+return(result_list)
+}
+
+
+#' Perform Linear Regression for GWAS
+#'
+#' @param genotype Numeric matrix of genotypes (n x p), where n is the number of samples and p is the number of SNPs.
+#' @param phenotype Numeric vector of phenotypes (length n).
+#' @param covariates Optional numeric matrix of covariates (n x k), where k is the number of covariates.
+#' @return A data frame containing regression results for each SNP, including \code{BETA}, \code{SE}, \code{Z}, \code{P}, and \code{Q}.
+#' @examples
+#' genotype <- matrix(rbinom(1000 * 10, 2, 0.3), 1000, 10)
+#' phenotype <- rnorm(1000)
+#' results <- run_linear_regression1(genotype, phenotype)
+run_linear_regression <- function(genotype, phenotype, covariates = NULL) {
+  if (!is.null(covariates)) {
+    covariates <- as.data.frame(lapply(covariates, as.numeric))
+  }
+  
+  reg_results <- univariate_regression(
+    X = genotype,
+    y = phenotype,
+    Z = covariates,
+    center = TRUE,
+    scale = FALSE
+  )
+  
+  snp_info <- lapply(colnames(genotype), parse_snp_info)
+
+  data.frame(
+    CHR = sapply(snp_info, function(x) x$chr),
+    BP = sapply(snp_info, function(x) x$pos),
+    SNP = colnames(genotype),
+    A1 = sapply(snp_info, function(x) x$alt),
+    A2 = sapply(snp_info, function(x) x$ref),
+    BETA = reg_results$betahat,
+    SE = reg_results$sebetahat,
+    Z = reg_results$z_scores,
+    P = reg_results$p_values,
+    Q = reg_results$q_values,
+    N = nrow(genotype)
+  )
+}
+
+
+#' Parse SNP name from chr:pos:ref:alt format
+#' @param snp character string in format "chr1:100:A:T"
+#' @return list with chr (numeric), pos (numeric), ref (character), alt (character)
+#' @examples
+#' parse_snp_info("chr1:100:A:T")
+parse_snp_info <- function(snp) {
```

Comment:
https://github.com/cumc/pecotmr/blob/main/R/misc.R#L278 these functions are similar. It's okay to have it in this format but let's move it to `misc.R`